GH-40028: [C++][FS][Azure] Add AzureFileSystem support to FileSystemFromUri()

kou opened the pull request:
### Rationale for this change

`FileSystemFromUri()` is a common API to create a file system object. `FileSystemFromUri()` should be able to create an `AzureFileSystem` object.

### What changes are included in this PR?

Add `AzureOptions::FromUri()` and use it from `FileSystemFromUri()`.

See the `AzureOptions::FromUri()`'s docstring about the supported formats.

### Are these changes tested?

Yes.

### Are there any user-facing changes?

Yes.
* GitHub Issue: #40028

kou commented:
@Tom-Newton Are `abfs://` and `abfss://` natural for `AzureFileSystem`?

Tom-Newton commented:
> @Tom-Newton Are `abfs://` and `abfss://` natural for `AzureFileSystem`?

I would say yes. I think originally they came from Hadoop with the extra "s" indicating secure but other filesystem implementations seem to have adopted it and they seem to be used mostly interchangeably.

nosterlu commented:
> > @Tom-Newton Are `abfs://` and `abfss://` natural for `AzureFileSystem`?
> 
> 
> 
> I would say yes. I think originally they came from Hadoop with the extra "s" indicating secure but other filesystem implementations seem to have adopted it and they seem to be used mostly interchangeably.

From the [documentation](https://learn.microsoft.com/en-us/azure/storage/blobs/data-lake-storage-introduction-abfs-uri) if it helps

'''
The abfs protocol is used as the scheme identifier. If you add an s at the end (abfss) then the ABFS Hadoop client driver will always use Transport Layer Security (TLS) irrespective of the authentication method chosen. If you choose OAuth as your authentication, then the client driver will always use TLS even if you specify abfs instead of abfss because OAuth solely relies on the TLS layer. Finally, if you choose to use the older method of storage account key, then the client driver interprets abfs to mean that you don't want to use TLS.
'''

kou commented:
Thanks for the note. I've added the documentation URL as a comment.

Tom-Newton commented:
> @kou @Tom-Newton can we use `azfs` (only `https` and by default) because this implementation uses both Blobs and File System APIs?

I have quite a strong opinion that we should support `abfss://` and `abfs://`. As a user I don't want to use different path formats when I use different filesystem implementations.

felipecrv commented:
> > @kou @Tom-Newton can we use `azfs` (only `https` and by default) because this implementation uses both Blobs and File System APIs?
> 
> I have quite a strong opinion that we should support the `abfss://` and `abfs://` Hadoop path formats. As a user I don't want to use different path formats when I use different filesystem implementations.

Fair enough, then what are the semantics of each URI format and how they map to *this* implementation?

I will start with one rule: both `abfss` and `abfs` map to `scheme=https` in `AzureOptions` cause people running pyarrow in insecure Wi-Fi networks shouldn't leak their credentials.

Tom-Newton commented:
> I will start with one rule: both `abfss` and `abfs` map to `scheme=https` in `AzureOptions`

That is fine with me. 

> what are the semantics of each URI format and how they map to this implementation?

For Hadoop format URIs I would probably just extract the storage account name. That means there is a lot of redundant information in the URL but I don't think that is really a problem and it gives us compatibility which I think is important. 

If we want to support other URIs formats too that would be useful to some people. Some examples I've seen on other filesystems: https://docs.rs/object_store/latest/object_store/azure/struct.MicrosoftAzureBuilder.html#method.with_url https://github.com/fsspec/adlfs

kou commented on cpp/src/arrow/filesystem/azurefs.cc:
```diff
@@ -65,6 +65,133 @@ AzureOptions::AzureOptions() = default;
+      options.blob_storage_scheme = kv.second;
+    } else if (kv.first == "dfs_storage_scheme") {
+      options.dfs_storage_scheme = kv.second;
+    } else if (kv.first == "credential_kind") {
```
Does it mean that we should use `ConfigureClientSecretCredential()` if `tenant_id`, `client_id` and `client_secret` are specified but `credential_kind=client_secret` isn't specified?

felipecrv replied:
`credential_kind` should never be specified and we should validate the URI to keep the invariant that it doesn't configure two different auth methods. And when nothing is provided, we use the default auth chain provided by the SDK.

kou replied:
Hmm, how can we distinguish `ConfigureAnonymousCredential()`, `ConfigureWorkloadIdentityCredential()` and `ConfigureDefaultCredential()`? All of them don't require additional information.

felipecrv replied:
The parameter-less auth methods can have dedicated query params for each. These being the valid configurations regarding auth:
- nothing (use default auth chain)
- `?anonymous`
- `?use_workload_identity`
- `?account_key=<ACCOUNT_KEY>`
- `?tenant_id=<TENANT_ID>&client_id=<CLIENT_ID>&client_secret=<CLIENT_SECRET>`
- `?client_id=<CLIENT_ID>` (client_id alone means managed identity credential)

kou replied:
`?anonymous` and `?use_workaround_identity` are conflicted parameters. (We can't specify both of them at once.) I think that it's better that we use the same parameter name for the type (`XXX={anonymous,workload_identity}`). If we use it, users can't specify both of them at once. (I know that URI spec accepts `XXX=anonymous&XXX=workload_identity`.)

How about accepting only (`default`, ) `anonymous` and `use_workload_identity` as valid `credential_kind` parameter?

- nothing (use default auth chain) -> nothing or `?credential_kind=default`
- `?anonymous` -> `?credential_kind=anonymous`
- `?use_workload_identity` -> `?credential_kind=workload_identity`
- `?account_key=<ACCOUNT_KEY>` -> not changed (`?credential_kind=storage_shared_key` is invalid)
- `?tenant_id=<TENANT_ID>&client_id=<CLIENT_ID>&client_secret=<CLIENT_SECRET>` -> not changed (`?credential_kind=client_secret` is invalid)
- `?client_id=<CLIENT_ID>` (client_id alone means managed identity credential) -> not changed (`?credential_kind=managed_identity` is invalid)

kou replied:
Ah, we don't need `?account_key=<ACCOUNT_KEY>` because we can get it from the URI's password part.

felipecrv replied:
> How about accepting only (default, ) anonymous and use_workload_identity as valid credential_kind parameter?

Sure. That looks good.

kou commented on cpp/src/arrow/filesystem/azurefs.h:
```diff
@@ -123,6 +126,13 @@ struct ARROW_EXPORT AzureOptions {
   ~AzureOptions();
 
+  /// Initialize from URIs such as
+  /// "abfs://account.blob.core.windows.net/container/dir/blob" and
+  /// "abfs://file_system@account.dfs.core.windows.net/dir/file".
```
We're discussing what URI is better for us in main comments: https://github.com/apache/arrow/pull/40325#issuecomment-1977200472

kou commented on cpp/src/arrow/filesystem/filesystem.cc:
```diff
@@ -690,6 +693,21 @@ Result<std::shared_ptr<FileSystem>> FileSystemFromUriReal(const Uri& uri,
     }
     return std::make_shared<LocalFileSystem>(options, io_context);
   }
+  /// "abfs" and "abfss" schemes are taken from
+  /// the Azure Data Lake Storage Gen2 URI:
+  /// https://learn.microsoft.com/en-us/azure/storage/blobs/data-lake-storage-introduction-abfs-uri
+  ///
+  /// abfs[s]://<file_system>@<account_name>.dfs.core.windows.net/<path>/<file_name>
```
Discussing on main comments: https://github.com/apache/arrow/pull/40325#issuecomment-1977200472

felipecrv commented:
> If we want to support other URIs formats too that would be useful to some people. Some examples I've seen on other filesystems: https://docs.rs/object_store/latest/object_store/azure/struct.MicrosoftAzureBuilder.html#method.with_url https://github.com/fsspec/adlfs

Thank you @Tom-Newton! This list from the Rust crate defines formats that allow us to express URIs that refer to the entire storage account and not just a specific filesystem. Plus, simple and short URIs that allow us to simply use the default endpoints. cc @kou

felipecrv commented on cpp/src/arrow/filesystem/azurefs.cc:
```diff
@@ -65,6 +65,133 @@ AzureOptions::AzureOptions() = default;
+    options.blob_storage_authority = host_port;
+    options.dfs_storage_authority = host_port;
+    if (uri.scheme() == "abfs") {
+      options.blob_storage_scheme = "http";
+      options.dfs_storage_scheme = "http";
```
So let's figure out a way to infer we're talking to Azurite, but not assume that people using `abfs://` are OK with plain-text communications and would remember to intentionally opt-in to security by using `abfss://`. Security should be opt-out, not opt-in. If someone ever complains about `https://` by default we should spec a URI query parameter to disable TLS instead of having it subtly disabled by the scheme chosen.

Even the Microsoft documentation says that they will use TLS even when `abfs` is used in the Haddop driver.

Tom-Newton replied:
Sounds good to me :+1:

felipecrv replied:
> We should process abfs and abfss as the same schemes
> We should always use HTTPS for abfs and abfss

Yes.

> We should add a query parameter to use HTTP (scheme=http?)

Probably something more explicit like `?disable_tls`. I would be OK with a change in `AzureOptions` as well. Instead of `scheme` a `bool disable_tls` and then we compute the scheme (http or https) according to it.

kou replied:
OK. Let's choose this approach.

I can understand `disable_tls` naming (it implies that TLS is enabled by default) but can we avoid using negative option name (`disable_XXX`)? It's a bit difficult to interpret when we use `if (!disable_XXX)` (double NOTs). Can we use `enable_tls=false` (the default value is `true` of cause)?

felipecrv replied:
Yeah. That's better.

kou commented:
URI list from https://docs.rs/object_store/latest/object_store/azure/struct.MicrosoftAzureBuilder.html#method.with_url :

* `abfs[s]://<container>/<path>` (according to [fsspec](https://github.com/fsspec/adlfs))
* `abfs[s]://<file_system>@<account_name>.dfs.core.windows.net/<path>`
* `abfs[s]://<file_system>@<account_name>.dfs.fabric.microsoft.com/<path>`
* `az://<container>/<path>`
* `adl://<container>/<path>`
* `azure://<container>/<path>`
* `https://<account>.dfs.core.windows.net`
* `https://<account>.blob.core.windows.net`
* `https://<account>.blob.core.windows.net/<container>`
* `https://<account>.dfs.fabric.microsoft.com`
* `https://<account>.dfs.fabric.microsoft.com/<container>`
* `https://<account>.blob.fabric.microsoft.com`
* `https://<account>.blob.fabric.microsoft.com/<container>`

We can use `abfs`/`abfss`/`az`/`adl`/`azure` schemes with the current `FileSystemFromUri()` mechanism. But `https` is a bit tricky. We need to check not only the scheme part but also the host part. This means that we need to put Azure related URI parsing code to `filesystem.cc` and `azurefs.cc`. And this isn't suitable for #39067 . #39067 only uses the scheme part to dispatch filesystem implementation.

felipecrv commented:
@kou let's implement only for now. Not even use the `az`, `adl`, `azure` schemes.

```
abfs[s]://<container>/<path> (according to [fsspec](https://github.com/fsspec/adlfs))
abfs[s]://<file_system>@<account_name>.dfs.core.windows.net/<path>
abfs[s]://<file_system>@<account_name>.dfs.fabric.microsoft.com/<path>
```

kou commented:
Can we also support one more format for Azurite?

```text
abfs[s]://<host>:<port>/<container>/<path>
```

If `<host>` doesn't have `.` and the `<port>` part doesn't exist, we will interpret the given URI as `abfs[s]://<container>/<path>`.

felipecrv commented:
> Can we also support one more format for Azurite?
> 
> ```
> abfs[s]://<host>:<port>/<container>/<path>
> ```
> 
> If `<host>` doesn't have `.` and the `<port>` part doesn't exist, we will interpret the given URI as `abfs[s]://<container>/<path>`.

Sure. I think that can work well.

kou commented:
OK. I'll implement the discussed spec.

kou commented:
Implemented:

Supported formats:

1. `abfs[s]://[:<password>@]<account>.blob.core.windows.net[/<container>[/<path>]]`
2. `abfs[s]://<container>[:<password>]@<account>.dfs.core.windows.net[/path]`
3. `abfs[s]://[<account[:<password>]@]<host[.domain]>[<:port>][/<container>[/path]]`
4. `abfs[s]://[<account[:<password>]@]<container>[/path]`

Added query parameters:

* `enable_tls`: It replaces `blob_storage_scheme` and `dfs_storage_scheme`
  parameters.

Removed query parameters:

* `blob_storage_scheme`: Replaced with `enable_tls`.
* `dfs_storage_scheme`: Replaced with `enable_tls`.

Changed query parameters:

* `credential_kind`: Accepts only `default`, `anonymous` and
  `workload_identity`.

kou commented:
I'll merge this in the next week if nobody objects it.
